### PR TITLE
clementine: fetchpatch to fix execution on wayland; some cleanup

### PR DESCRIPTION
**Pull request by evils**
###### Motivation for this change
wanted to try out `clementine`, it segfaulted at startup

###### Things done
`nix-update clementine --version=unstable --commit`
`nixpkgs-fmt`
add some dependencies the config phase noted as not found ¯\\\_(ツ)\_/¯

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - `/nix/store/6py3kv649fdyx2i11qn6vj9pjf4xfq04-clementine-free-1.4.0rc1    915311856`
  - `/nix/store/yawqv0bl36q6scnn0sx23lxd6x2fvhhv-clementine-free-1.4.0rc1    915321680`
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121493` at b0e13961 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>clementine</li>
    <li>clementineUnfree</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/applications/audio/clementine/default.nix:1:39:
```
  |
1 | { lib, mkDerivation, fetchFromGitHub, fetchpatch, boost, cmake, chromaprint, gettext, gst_all_1, liblastfm
  |                                       ^
```
</li>
  </ul>
</details>

**Comment by evils** [comment]
@ttuegel seems likely you missed this

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 121493` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>clementine</li>
    <li>clementineUnfree</li>
  </ul>
</details>

**Comment by timokau** [comment]
This would move our clementine version from "release candidate" to "nightly" territory. I would prefer to avoid that if we can. Have you looked into the source of the error? Would your first commit (adding the missing dependencies) maybe fix it on its own?

If it does not take too long to test a build on your machine, you could also try to bisect the issue. You can clone the clementine project, start a bisect (the meaning of good/bad are exchanged here, so you might want to use different terms to avoid confusion) and simply use the revisions the bisection gives you to declare the `src` attribute. You could also replace the `src = ...` line to directly build from your local checkout.

**Review by timokau — CHANGES_REQUESTED** [review]
Needs info, see my previous comment.

**Comment by evils** [comment]
> This would move our clementine version from "release candidate" to "nightly" territory.

fair point, i hadn't considered this as a significant change

their RC1 is from 2020-01-02 and they've had 604 commits since then, it doesn't seem unreasonable to take a CI/CD pre-release at this point...
(though bumping this PR to the latest before it is merged seems like a good idea to me)

the segfault occurs in `XKeysymToKeycode` from `libX11.so.6`, and appears to be specific to wayland

there is an upstream issue open for this
https://github.com/clementine-player/Clementine/issues/6587
whence there is a "short-term fix" whose change appears to still be in the master branch, via `2179027a6d97530c857e43be873baacd696ff332`
i suppose we could fetchpatch that one, but that feels like leaving 603 improvements on the table to me

this PR was only ever intended to be a courtesy side-effect of my trying out of the app, i wasn't expecting complications (though you raise a valid point) and would have hoped the maintainer would take it from there...

**Comment by timokau** [comment]
I understand that it can be frustrating if a drive-by contribution drags on for too long and takes more effort than expected, but at the same time I do not want to compromise on review standards. Thank you for going the extra mile and looking for the source of the error.

I would prefer the `fetchpatch` solution for multiple reasons:

- it's upstream's choice when to cut a release. "Overriding" them on this can have negative consequences, like nix users filing upstream issues while not being aware that they are using an unstable version. See [this](https://news.ycombinator.com/item?id=19936320) related comment.
- 603 improvements are also 603 opportunities for new bugs. I don't know upstream's policy on changes to master and how "releasable" it is at any point in time. Again, I think its best to rely on them to decide which revisions end users should use.
- Shipping an unstable version is against the user's expectations. Users will likely assume that they are using the latest (or something close to the latest) stable version, which can be confusing e.g. when looking up help or documentation online.

I'm not opposed to shipping unstable versions in general, but if we do I would prefer if we are very explicit about this (i.e. adding an `-unstable` postfix to the package name or similar).

Would you consider staying on the old version and using `fetchpatch` instead? I'm not sure about the reformatting commit (there was some recent discussion about applying a formatter to nixpkgs, but I'm not sure what the current consensus is). Maybe it would be better to wait for a decision on treewide formatter, but I won't block this based on the formatting commit either way.

**Comment by evils** [comment]
> I understand that it can be frustrating if a drive-by contribution drags on for too long

i'm more annoyed at seeing the maintainer merge another (newer) PR for this package but not even comment on this one

> I do not want to compromise on review standards.

please don't

> Would you consider staying on the old version and using fetchpatch instead?

i did bring that up as an option, i was just a bit hesitant because 603 other commits sounds like a lot of opportunity for that patch to have unintended consequences, but it seems simple enough
i don't think this is up to me to decide, but i wasn't going to do so without at least mentioning how far behind the rc1 is

> I'm not sure about the reformatting commit

i found it necessary to reformat at least the `nativeBuildInputs` part (the additions to which i made purely to suppress some configuration "not found" warnings),
at which point i preferred to use a tool to do so

(bikeshedding side note: i like having some freedom in package formatting,
like having a few short inputs on the same line,
or having a closing `;` on a separate line in a package that gets refactored often
or having parenthesis around a single logical statement in an `if` if there's a chance i'd want to change the content
and think at least `nixpkgs-fmt` would be too strict in that regard)

**Comment by evils** [comment]
just mentioning it in a comment as i don't think anyone gets notified of commit and title changes

this now uses the RC1 plus the wayland fix commit

plus the cleanup i had previously (i confirmed the missing config stuff is still looked for)

File: pkgs/applications/audio/clementine/default.nix
Comment (timokau) [inline pkgs/applications/audio/clementine/default.nix:66]: Please add a reference to the upstream issue as well. It's nice to have that context if someone wants to look into the background of the patch.

**Review by timokau — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/clementine/default.nix
Comment (evils) [reply pkgs/applications/audio/clementine/default.nix:66]: right!

**Comment by timokau** [comment]
> > I understand that it can be frustrating if a drive-by contribution drags on for too long
> 
> i'm more annoyed at seeing the maintainer merge another (newer) PR for this package but not even comment on this one

I understand your frustration, but at the same time I also very much can emphasize that sometimes it's hard to find the time to respond to all open-source related notifications. It is volunteer work after all.

> i was just a bit hesitant because 603 other commits sounds like a lot of opportunity for that patch to have unintended consequences, but it seems simple enough

That is also a valid point, but I would guess that the "breakage potential" of such a small patch is still smaller than that of an update to a nightly version.

> (bikeshedding side note: i like having some freedom in package formatting,
> like having a few short inputs on the same line,
> or having a closing `;` on a separate line in a package that gets refactored often
> or having parenthesis around a single logical statement in an `if` if there's a chance i'd want to change the content
> and think at least `nixpkgs-fmt` would be too strict in that regard)

I think the consistent formatting and "taking formatting off the table" during reviews would be worth it. Using formatters on parts of the tree without consensus has potential downsides (may happen multiple times with different formatters, makes git history a bit harder to use, increases diff size), but either way the changes here should be good to go if you add the reference to the upstream wayland issue.

File: pkgs/applications/audio/clementine/default.nix
Comment (evils) [reply pkgs/applications/audio/clementine/default.nix]: done

**Review by timokau — APPROVED** [review]
Thank you for taking the time to fix this, even though you may not be a regular clementine user.

**Pull request merged**
